instrumentation/cassandra/cassandra-4.0/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/cassandra/v4_0/CassandraDatabaseClientTracer.java

```diff
@@ -49,14 +62,85 @@ protected InetSocketAddress peerAddress(CqlSession cqlSession) {
     return null;
   }
 
+  @Override
+  protected Span onConnection(Span span, CqlSession cqlSession) {
+    span = super.onConnection(span, cqlSession);
+    DriverExecutionProfile config = cqlSession.getContext().getConfig().getDefaultProfile();
+    // may be overwritten by statement, but take the default for now
+    int pageSize = config.getInt(DefaultDriverOption.REQUEST_PAGE_SIZE);
+    if (pageSize > 0) {
+      span.setAttribute(SemanticAttributes.DB_CASSANDRA_PAGE_SIZE, pageSize);
+    }
+    // may be overwritten by statement, but take the default for now
+    span.setAttribute(
+        SemanticAttributes.DB_CASSANDRA_CONSISTENCY_LEVEL,
+        config.getString(DefaultDriverOption.REQUEST_CONSISTENCY));
+    return span;
+  }
+
   public void onResponse(Context context, ExecutionInfo executionInfo) {
+    Span span = Span.fromContext(context);
     Node coordinator = executionInfo.getCoordinator();
     if (coordinator != null) {
       SocketAddress socketAddress = coordinator.getEndPoint().resolve();
       if (socketAddress instanceof InetSocketAddress) {
-        Span span = Span.fromContext(context);
         NetPeerUtils.INSTANCE.setNetPeer(span, ((InetSocketAddress) socketAddress));
       }
+      if (coordinator.getDatacenter() != null) {
+        span.setAttribute(
+            SemanticAttributes.DB_CASSANDRA_COORDINATOR_DC, coordinator.getDatacenter());
+      }
+      if (coordinator.getHostId() != null) {
+        span.setAttribute(
+            SemanticAttributes.DB_CASSANDRA_COORDINATOR_ID, coordinator.getHostId().toString());
+      }
+    }
+    span.setAttribute(
+        SemanticAttributes.DB_CASSANDRA_SPECULATIVE_EXECUTION_COUNT,
+        executionInfo.getSpeculativeExecutionCount());
+
+    Statement<?> statement = executionInfo.getStatement();
+    // override connection default if present
+    if (statement.getConsistencyLevel() != null) {
+      span.setAttribute(
+          SemanticAttributes.DB_CASSANDRA_CONSISTENCY_LEVEL,
+          statement.getConsistencyLevel().name());
+    }
+    // override connection default if present
+    if (statement.getPageSize() > 0) {
+      span.setAttribute(SemanticAttributes.DB_CASSANDRA_PAGE_SIZE, statement.getPageSize());
+    }
+    span.setAttribute(SemanticAttributes.DB_CASSANDRA_IDEMPOTENCE, isIdempotent(statement));
+  }
+
+  @Override
+  protected void onStatement(Span span, String statement) {
+    super.onStatement(span, statement);
+    String table = extractTableNameFromQuery(statement);
+    if (table != null) {
+      span.setAttribute(SemanticAttributes.DB_CASSANDRA_TABLE, table);
+    }
+  }
+
+  @Nullable
+  private static String extractTableNameFromQuery(String query) {
+    String tableName = null;
+    Matcher matcher = tableNameRegex.matcher(query);
+    if (matcher.find()) {
+      if (matcher.group(2) != null) {
```

Comment:
Do we have a unit test for these two types of matches?

Reply:
I have added unit tests

Reply:
@anuraaga I can't figure out why the unit test is unable to see the public CassandraTableNameExtractor (and before that, the embedded extractor method). Am I missing some test directives in the gradle file?

Reply:
for ref the task is :instrumentation:cassandra:cassandra-4.0:javaagent:test

Reply:
Sorry for the trouble - we merged a change which has our tests behave more like an instrumented app, meaning indeed they don't get access to the agent's classes. You'll need to split out a unit tests module for them (long term we should split out library instrumentation instead).

https://github.com/open-telemetry/opentelemetry-java-instrumentation/tree/master/instrumentation/jdbc/javaagent-unittests

Reply:
should be good now @anuraaga